src/peft/utils/other.py

```diff
@@ -95,10 +95,10 @@ def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
         param.requires_grad = False
 
     if not is_gptq_quantized:
-        # cast all non INT8 parameters to fp32
+        # cast all non INT8 parameters to torch_dtype
         for param in model.parameters():
-            if (param.dtype == torch.float16) or (param.dtype == torch.bfloat16):
-                param.data = param.data.to(torch.float32)
+            if param.dtype not in [torch.int8, torch.uint8]:
```

Comment:
Does this check work with 4-bit/3-bit quantization?

Reply:
@pacman100. Thanks for the info. Indeed, setting to fp16  will lead to  Nan and loss 0, and there is not any warning or error. 

I spent a few hours trying to find out why. I am using llama2 for lora finetuning, using peft+accelerate(+deepspeed) here is what I found out, for now:
- overflow is found in the first forward pass of my training process,  not even reaching the backward pass
-  bnb is OK doing its job.
- RMSNorm for llama is OK, as it will  cast input to float32 and back.
- attention is OK , 4096 x 4096 seems ok for my test
- Overflow found in MLP, due to the up projection matrix multiplication, 11008x4096 matrix.

Reply:
4-bit is stored as torch.uint8 as I know.  but I a have no idea about 3bit. ?